model-optimizer/extensions/back/add_outputs_recursive.py

```diff
@@ -0,0 +1,259 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+import numpy as np
+
+from extensions.ops.loop import Loop
+from extensions.ops.tensor_iterator import TensorIterator
+from mo.back.replacement import BackReplacementPattern
+from mo.front.common.partial_infer.utils import int64_array
+from mo.front.tf.graph_utils import create_op_node_with_second_input
+from mo.graph.graph import Graph, Node
+from mo.ops.result import Result
+from mo.ops.unsqueeze import Unsqueeze
+
+
+# shape inference for TensorIterator
+# copy shapes from internal nodes + insert correct iterations count where needed
+def ti_infer(step_node, port_num):
```

Comment:
Maybe the function `ti_infer(step_node, port_num)` should be rewritten using less nesting levels? For example, as
```python
def ti_infer(step_node, port_num):
    out_port_map = step_node.output_port_map
    int_layer_id = None
    port_num = port_num + len(step_node.in_ports())
    # find out iterations count for TensorIterator to set output shape correctly
    iterations_count = 1
    found_om = None
    for om in out_port_map:
        if om['external_port_id'] == port_num:
            found_om = om
            break
    if found_om is not None:
        int_layer_id = om['internal_layer_id']
        if 'axis' in om and om['axis'] is not None and 'end' in om and om['end'] is not None:
            if om['end'] != -1:
                iterations_count = (om['end'] - om['start']) / om['stride']
            else:
                for in_rec in step_node.input_port_map:
                    if 'axis' in in_rec and in_rec['axis'] is not None:
                        if in_rec['end'] != -1:
                            in_shape_axis = in_rec['end']
                        else:
                            in_shape_axis = step_node.in_port(in_rec['external_port_id']).data.get_shape()[
                                in_rec['axis']]
                        iterations_count = (in_shape_axis - in_rec['start']) / in_rec['stride']
                        break
    int_node_name = TensorIterator.find_internal_layer_id(step_node.body, int_layer_id)
    int_node = Node(step_node.body, int_node_name)
    assert int_node.op == 'Result'
    out_shape = int_node.in_port(0).data.get_shape().copy()
    port_num = port_num - len(step_node.in_ports())

    out_shape[0] = iterations_count

    step_node.out_port(port_num).data.set_shape(out_shape)
```
Because actually this function finds `om` in `out_port_map` (with `om['external_port_id'] == port_num`) and performs some actions with found `om`.

Reply:
agree, rewritten code in more compact manner